Sources/SwiftSyntax/SyntaxText.swift

```diff
@@ -0,0 +1,226 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#if canImport(Darwin)
+import Darwin
+#elseif canImport(Glibc)
+import Glibc
+#endif
+
+/// Represent a string.
+///
+/// This type does not own the string data. The data reside in some other buffer
+/// whose lifetime extends past that of the SyntaxText.
+///
+/// `SyntaxText` conforms to `Collection` where the element is `UInt8`.
+@_spi(Testing) // SPI name is subject to change
+public struct SyntaxText {
+  var buffer: UnsafeBufferPointer<UInt8>
+
+  public init(baseAddress: UnsafePointer<UInt8>?, count: Int) {
+    assert(count == 0 || baseAddress != nil,
+           "If count is not zero, base address must be exist")
+    buffer = .init(start: baseAddress, count: count)
+  }
+
+  /// Creates an empty `SyntaxText`
+  public init() {
+    self.init(baseAddress: nil, count: 0)
+  }
+
+  /// Creates a `SyntaxText` from a `StaticString`
+  public init(_ string: StaticString) {
+    self.init(baseAddress: string.utf8Start, count: string.utf8CodeUnitCount)
+  }
+
+  /// Base address of the memory range this string refers to.
+  public var baseAddress: UnsafePointer<UInt8>? {
+    buffer.baseAddress
+  }
+
+  /// Byte length of this string.
+  public var count: Int {
+    buffer.count
+  }
+
+  /// A Boolean value indicating whether a string has no characters.
+  public var isEmpty: Bool {
+    buffer.isEmpty
+  }
+
+  /// Returns `true` if the memory range of this string is a part of `other`.
+  ///
+  /// `text[n ..< m].isSliceOf(text)` is always true as long as `n` and `m` are
+  /// valid indices.
+  public func isSliceOf(_ other: SyntaxText) -> Bool {
+    guard !isEmpty && !other.isEmpty else {
+      return isEmpty == other.isEmpty
+    }
+    return (other.baseAddress! <= baseAddress! &&
+            baseAddress! + count <= other.baseAddress! + other.count)
+  }
+
+  /// Returns `true` if `other` is a substring of this `SyntaxText`.
+  ///
+  /// This should not be used because `SyntaxText` is a `Collection<UInt8>` and
+  /// `contains(_:)` is usually for checking a collection contains the element.
+  /// Using `contains(_:)` for subsequence is confusing.
+  @available(*, deprecated, message: "Use 'firstRange(of:) != nil'")
+  public func contains(_ other: SyntaxText) -> Bool {
+    return firstRange(of: other) != nil
+  }
+
+  /// Finds and returns the range of the first occurrence of `other` within this
+  /// string. Returns `nil` if `other` is not found.
+  public func firstRange(of other: SyntaxText) -> Range<Index>? {
+    if other.isEmpty { return nil }
```

Comment:
Don’t we also need to check that `self` is not empty?

```suggestion
    if isEmpty || other.isEmpty { return nil }
```

Reply:
I will, because I will so `let selfBaseAddress = self.baseAddress!` outside the loop, but it was not needed because `self` was empty, `while` condition is never satisfied.

Reply:
Since I stoped hoisting optional unwrapping. I left this as-is. with some code comments.

Reply:
Makes sense 👍